pkg/container-collection/options.go

```diff
@@ -0,0 +1,386 @@
+				&pb.ContainerDefinition{
+					Id:  container.ID,
+					Pid: uint32(pid),
+				})
+		}
+
+		// Future containers
+		cc.containerEnrichers = append(cc.containerEnrichers, func(container *pb.ContainerDefinition) bool {
+			filter := dockerfilters.NewArgs()
+			filter.Add("id", container.Id)
+			containers, err := dockercli.ContainerList(context.Background(),
+				dockertypes.ContainerListOptions{
+					All:     true,
+					Filters: filter,
+				})
+			if err != nil {
+				log.Errorf("failed to list container %s: %s", container.Id, err)
+				return true
+			}
+			if len(containers) == 1 {
+				if len(containers[0].Names) > 0 {
+					container.Podname = strings.TrimPrefix(containers[0].Names[0], "/")
+				}
+			} else {
+				log.Errorf("container %s has %d names", container.Id, len(containers))
+			}
+			return true
+		})
+		return nil
+	}
+}
+
+// WithPodInformer uses a pod informer to get both initial containers and the
+// stream of container events. It then uses the CRI interface to get the
+// process ID.
+//
+// This cannot be used together with WithInitialKubernetesContainers() since
+// the pod informer already gets initial containers.
+func WithPodInformer(nodeName string) ContainerCollectionOption {
+	return func(cc *ContainerCollection) error {
+		k8sClient, err := k8s.NewK8sClient(nodeName)
+		if err != nil {
+			return fmt.Errorf("failed to create Kubernetes client: %w", err)
+		}
+
+		createdChan := make(chan *v1.Pod)
+		deletedChan := make(chan string)
+
+		_, err = k8s.NewPodInformer(nodeName, createdChan, deletedChan)
+		if err != nil {
+			return fmt.Errorf("failed to create pod informer: %w", err)
+		}
+
+		go func() {
+			// containerIDsByKey keeps track of container ids for each key. This is
+			// necessary because messages from deletedChan only gives the key
+			// without additional context.
+			//
+			// key is "namespace/podname"
+			// value is an set of containerId
+			containerIDsByKey := make(map[string]map[string]struct{})
+
+			for {
+				select {
+				case d := <-deletedChan:
+					if containerIDs, ok := containerIDsByKey[d]; ok {
+						for containerID := range containerIDs {
+							cc.RemoveContainer(containerID)
+						}
+					}
+				case c := <-createdChan:
+					containers := k8sClient.PodToContainers(c)
+					key, _ := cache.MetaNamespaceKeyFunc(c)
+					containerIDs, ok := containerIDsByKey[key]
+					if !ok {
+						containerIDs = make(map[string]struct{})
+						containerIDsByKey[key] = containerIDs
+					}
+					for _, container := range containers {
+						// The container is already registered, there is not any chance the
+						// PID will change, so ignore it.
+						if _, ok := containerIDs[container.Id]; ok {
+							continue
+						}
+
+						// Make a copy instead of passing the same pointer at
+						// each iteration of the loop
+						newContainer := pb.ContainerDefinition{}
+						newContainer = container
+
+						cc.AddContainer(&newContainer)
```

Comment:
I had this initially, and it adds the same container multiple times in the `sync.Map` (`gadgettracermanager -dump` would show the same container duplicated). This is because the same pointer is added multiple times, and at each iteration of the loop, the loop variable is updated, so that's updating the others containers already added in the `sync.Map`.